thorlcr/thorutil/thmem.cpp

```diff
@@ -1679,31 +1686,70 @@ class CThorRowCollectorBase : public CSpillable
     }
     IRowStream *getStream(CThorExpandingRowArray *allMemRows, memsize_t *memUsage, bool shared)
     {
-        CriticalBlock b(readerLock);
-        if (0 == outStreams)
         {
-            flush();
-            if (spillingEnabled())
+            CThorArrayLockBlock block(spillableRows); // ensure locked until deactivated
+            if (0 == outStreams++)
             {
-                // i.e. all disk OR (some on disk already AND allDiskOrAllMem)
-                if (((rc_allDisk == diskMemMix) || ((rc_allDiskOrAllMem == diskMemMix) && overflowCount)))
+                flush();
+                if (spillingEnabled())
                 {
-                    CThorArrayLockBlock block(spillableRows);
-                    if (spillableRows.numCommitted())
+                    // i.e. all disk OR (some on disk already AND allDiskOrAllMem)
+                    if (((rc_allDisk == diskMemMix) || ((rc_allDiskOrAllMem == diskMemMix) && overflowCount)))
                     {
-                        spillRows(false);
-                        spillableRows.kill();
+                        if (spillableRows.numCommitted())
+                        {
+                            spillRows(false);
+                            spillableRows.kill();
+                        }
                     }
                 }
+
+                /* Ensure existing callback is cleared, before:
+                 * a) instrms are built, since new spillFiles can't be added to as long as existing callback is active
+                 * b) streams created based on spillableRows, which take ownership of spillableRows and in turn add their own callbacks
+                 */
+                deactivateSpillingCallback(); // NB: spillableRows can no longer be altered asynchronously
+
+                if (spillableRows.numCommitted())
+                {
+                    totalRows += spillableRows.numCommitted();
+                    if (iCompare)
+                    {
+                        // Option(rcflag_noAllInMemSort) - avoid sorting allMemRows
```

Comment:
Not  connected with this PR, but I don't think this option is ever used (I was trying to work out why it would be used...)

Reply:
No, I can see it's not. And although not directly related, I am going to remove associated code now.